pkg/actuators/machine/utils.go

```diff
@@ -277,3 +278,50 @@ func findAWSMachineProviderCondition(conditions []providerconfigv1.AWSMachinePro
 	}
 	return nil
 }
+
+// awsInstanceRegMatch represents Regex Match for AWS instance.
+var awsInstanceRegMatch = regexp.MustCompile("^i-[^/]*$")
+
+// mapToAWSInstanceID extracts the InstanceID from the KubernetesInstanceID
+// providerID represents the id for an instance in the kubernetes API;
+// the following form
+//  * aws:///<zone>/<awsInstanceId>
+//  * aws:////<awsInstanceId>
+//  * <awsInstanceId>
+func mapToAWSInstanceID(providerID string) (*string, error) {
+	if !strings.HasPrefix(providerID, "aws://") {
+		// Assume a bare aws volume id (vol-1234...)
+		// Build a URL with an empty host (AZ)
+		providerID = "aws://" + "/" + "/" + providerID
+	}
+	url, err := url.Parse(providerID)
+	if err != nil {
+		return nil, fmt.Errorf("invalid provider id (%q): %v", providerID, err)
+	}
+	if url.Scheme != "aws" {
+		return nil, fmt.Errorf("invalid scheme for AWS instance (%q)", providerID)
+	}
+
+	// awsID represents the ID of the instance in the AWS API, e.g. i-12345678
+	// The "traditional" format is "i-12345678"
+	// A new longer format is also being introduced: "i-12345678abcdef01"
+	// We should not assume anything about the length or format, though it seems
+	// reasonable to assume that instances will continue to start with "i-".
+	awsID := ""
+	tokens := strings.Split(strings.Trim(url.Path, "/"), "/")
+	if len(tokens) == 1 {
```

Comment:
This makes even `aws:////////////<awsInstanceId>` a valid case. Is that intended? If you know we always expect `aws://` prefix, you can put entire `providerID` string under a single regexp. Or, make sure you count the number of backslashes.

Reply:
updated!